### ISSUE TITLE
Cannot install dependencies with npm 3

### ISSUE THREAD

**Issue by Tyriar** [Microsoft/vscode#2474]
I'm hitting the following when installing deps (Ubuntu 15.10). `extensions/vscode-api-tests/node_modules/vscode/bin/install` doesn't seem to exist, only `compile` exists in that directory.

```
~/dev/forks/vscode @master
❯ git clean -fd && rm -rf node_modules out

~/dev/forks/vscode @master
❯ ./scripts/npm.sh install --arch=x64
npm WARN deprecated lodash@1.0.2: lodash@<2.0.0 is no longer maintained. Upgrade to lodash@^3.0.0
loadDep:minimatch → 304   ▐ ╢█████████████████████████░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░╟
WARN engine prettyjson@0.8.1: wanted: {"node":">= 0.6.0 < 0.11.0"} (current: {"node":"4.2.0","npm":"3.5.npm WARN deprecated lodash@0.9.2: lodash@<2.0.0 is no longer maintained. Upgrade to lodash@^3.0.0
npm WARN deprecated has-color@0.1.7: Renamed to supports-color. If you're using chalk, upgrade to the latest version. https://github.com/chalk/supports-color

> typechecker@2.0.8 preinstall /home/daniel/dev/forks/vscode/node_modules/.staging/typechecker-99c82ac6cec4a40884b7a49043a3e497
> node ./cyclic.js

npm WARN prefer global js-beautify@1.5.10 should be installed with -g

> fsevents@0.3.8 install /home/daniel/dev/forks/vscode/node_modules/chokidar/node_modules/fsevents
> node-gyp rebuild

make: Entering directory '/home/daniel/dev/forks/vscode/node_modules/chokidar/node_modules/fsevents/build'
  SOLINK_MODULE(target) Release/obj.target/.node
  COPY Release/.node
make: Leaving directory '/home/daniel/dev/forks/vscode/node_modules/chokidar/node_modules/fsevents/build'

> ghooks@1.0.3 install /home/daniel/dev/forks/vscode/node_modules/ghooks
> node ./bin/install


> alexandrudima-oniguruma@6.0.1 install /home/daniel/dev/forks/vscode/node_modules/vscode-textmate/node_modules/alexandrudima-oniguruma
> node-gyp rebuild

make: Entering directory '/home/daniel/dev/forks/vscode/node_modules/vscode-textmate/node_modules/alexandrudima-oniguruma/build'
  CC(target) Release/obj.target/oniguruma/deps/onig/regcomp.o
  CC(target) Release/obj.target/oniguruma/deps/onig/regenc.o
  CC(target) Release/obj.target/oniguruma/deps/onig/regerror.o
  CC(target) Release/obj.target/oniguruma/deps/onig/regexec.o
  CC(target) Release/obj.target/oniguruma/deps/onig/regext.o
  CC(target) Release/obj.target/oniguruma/deps/onig/reggnu.o
  CC(target) Release/obj.target/oniguruma/deps/onig/regparse.o
  CC(target) Release/obj.target/oniguruma/deps/onig/regposerr.o
  CC(target) Release/obj.target/oniguruma/deps/onig/regposix.o
  CC(target) Release/obj.target/oniguruma/deps/onig/regsyntax.o
  CC(target) Release/obj.target/oniguruma/deps/onig/regtrav.o
  CC(target) Release/obj.target/oniguruma/deps/onig/regversion.o
  CC(target) Release/obj.target/oniguruma/deps/onig/st.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/ascii.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/big5.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/cp1251.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/euc_jp.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/euc_kr.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/euc_tw.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/gb18030.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/iso8859_1.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/iso8859_2.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/iso8859_3.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/iso8859_4.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/iso8859_5.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/iso8859_6.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/iso8859_7.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/iso8859_8.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/iso8859_9.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/iso8859_10.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/iso8859_11.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/iso8859_13.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/iso8859_14.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/iso8859_15.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/iso8859_16.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/koi8.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/koi8_r.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/mktable.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/sjis.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/unicode.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/utf16_be.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/utf16_le.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/utf32_be.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/utf32_le.o
  CC(target) Release/obj.target/oniguruma/deps/onig/enc/utf8.o
  AR(target) Release/obj.target/oniguruma.a
  COPY Release/oniguruma.a
  CXX(target) Release/obj.target/onig_scanner/src/onig-result.o
  CXX(target) Release/obj.target/onig_scanner/src/onig-reg-exp.o
  CXX(target) Release/obj.target/onig_scanner/src/onig-scanner.o
  CXX(target) Release/obj.target/onig_scanner/src/onig-scanner-worker.o
  CXX(target) Release/obj.target/onig_scanner/src/onig-searcher.o
  CXX(target) Release/obj.target/onig_scanner/src/onig-string.o
  SOLINK_MODULE(target) Release/obj.target/onig_scanner.node
  COPY Release/onig_scanner.node
make: Leaving directory '/home/daniel/dev/forks/vscode/node_modules/vscode-textmate/node_modules/alexandrudima-oniguruma/build'

> native-keymap@0.1.2 install /home/daniel/dev/forks/vscode/node_modules/native-keymap
> node-gyp rebuild

make: Entering directory '/home/daniel/dev/forks/vscode/node_modules/native-keymap/build'
  CXX(target) Release/obj.target/keymapping/src/string_conversion.o
  CXX(target) Release/obj.target/keymapping/src/keymapping.o
  CXX(target) Release/obj.target/keymapping/deps/chromium/x/keysym_to_unicode.o
  CXX(target) Release/obj.target/keymapping/src/keyboard_x.o
  SOLINK_MODULE(target) Release/obj.target/keymapping.node
  COPY Release/keymapping.node
make: Leaving directory '/home/daniel/dev/forks/vscode/node_modules/native-keymap/build'

> weak@1.0.1 install /home/daniel/dev/forks/vscode/node_modules/weak
> node-gyp rebuild

make: Entering directory '/home/daniel/dev/forks/vscode/node_modules/weak/build'
  CXX(target) Release/obj.target/weakref/src/weakref.o
  SOLINK_MODULE(target) Release/obj.target/weakref.node
  COPY Release/weakref.node
make: Leaving directory '/home/daniel/dev/forks/vscode/node_modules/weak/build'

> windows-mutex@0.2.0 install /home/daniel/dev/forks/vscode/node_modules/windows-mutex
> node-gyp rebuild

make: Entering directory '/home/daniel/dev/forks/vscode/node_modules/windows-mutex/build'
  CXX(target) Release/obj.target/CreateMutex/src/main.o
../src/main.cc:2:21: fatal error: windows.h: No such file or directory
 #include <windows.h>
                     ^
compilation terminated.
CreateMutex.target.mk:91: recipe for target 'Release/obj.target/CreateMutex/src/main.o' failed
make: *** [Release/obj.target/CreateMutex/src/main.o] Error 1
make: Leaving directory '/home/daniel/dev/forks/vscode/node_modules/windows-mutex/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/home/daniel/.npm-packages/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:276:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Linux 4.2.0-23-generic
gyp ERR! command "/usr/local/bin/node" "/home/daniel/.npm-packages/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /home/daniel/dev/forks/vscode/node_modules/windows-mutex
gyp ERR! node -v v4.2.0
gyp ERR! node-gyp -v v3.2.0
gyp ERR! not ok 
npm WARN install:windows-mutex@0.2.0 windows-mutex@0.2.0 install: `node-gyp rebuild`
npm WARN install:windows-mutex@0.2.0 Exit status 1

> Code@0.10.7 preinstall /home/daniel/dev/forks/vscode
> node build/npm/preinstall.js


> Code@0.10.7 postinstall /home/daniel/dev/forks/vscode
> npm --prefix extensions/vscode-api-tests/ install extensions/vscode-api-tests/ && npm --prefix extensions/json/ install extensions/json/


> vscode-api-tests@0.0.1 postinstall /home/daniel/dev/forks/vscode/extensions/vscode-api-tests
> node ./node_modules/vscode/bin/install

module.js:339
    throw err;
    ^

Error: Cannot find module '/home/daniel/dev/forks/vscode/extensions/vscode-api-tests/node_modules/vscode/bin/install'
    at Function.Module._resolveFilename (module.js:337:15)
    at Function.Module._load (module.js:287:25)
    at Function.Module.runMain (module.js:467:10)
    at startup (node.js:134:18)
    at node.js:961:3

npm ERR! Linux 4.2.0-23-generic
npm ERR! argv "/usr/local/bin/node" "/home/daniel/.npm-packages/bin/npm" "--prefix" "extensions/vscode-api-tests/" "install" "extensions/vscode-api-tests/"
npm ERR! node v4.2.0
npm ERR! npm  v3.5.1
npm ERR! code ELIFECYCLE
npm ERR! vscode-api-tests@0.0.1 postinstall: `node ./node_modules/vscode/bin/install`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the vscode-api-tests@0.0.1 postinstall script 'node ./node_modules/vscode/bin/install'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the vscode-api-tests package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node ./node_modules/vscode/bin/install
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs vscode-api-tests
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls vscode-api-tests
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/daniel/dev/forks/vscode/npm-debug.log

npm ERR! Linux 4.2.0-23-generic
npm ERR! argv "/usr/local/bin/node" "/home/daniel/.npm-packages/bin/npm" "install" "--arch=x64"
npm ERR! node v4.2.0
npm ERR! npm  v3.5.1
npm ERR! code ELIFECYCLE
npm ERR! Code@0.10.7 postinstall: `npm --prefix extensions/vscode-api-tests/ install extensions/vscode-api-tests/ && npm --prefix extensions/json/ install extensions/json/`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the Code@0.10.7 postinstall script 'npm --prefix extensions/vscode-api-tests/ install extensions/vscode-api-tests/ && npm --prefix extensions/json/ install extensions/json/'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the Code package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     npm --prefix extensions/vscode-api-tests/ install extensions/vscode-api-tests/ && npm --prefix extensions/json/ install extensions/json/
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs Code
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls Code
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/daniel/dev/forks/vscode/npm-debug.log
```

**Comment by Tyriar** [comment]
Removing `"postinstall": "node ./node_modules/vscode/bin/install"` from `extensions/vscode-api-tests` locally gets past this error.

**Comment by joaomoreno** [comment]
@bpasero maybe related to npm 3?

**Comment by bpasero** [comment]
Yeah, npm@2 works. 

@Tyriar if you can find out why it fails, please ping me. Maybe there is a verbose option that prints why npm 3 is stupid and 2 is not. The related package.json entry that causes this is:

npm --prefix extensions/vscode-api-tests/ install extensions/vscode-api-tests/

**Comment by Tyriar** [comment]
It's the post install trying to run `extensions/vscode-api-tests/node_modules/vscode/bin/install`. 

Is this file generated by something? I managed to run the app tests by pulling this postinstall step out completely.

**Comment by bpasero** [comment]
@Tyriar no it is there, use npm 2 and everything just works fine. are you saying you dont have this file in vscode-api-tests? what version of vscode module do you have in vscode-api-tests/node_modules?

**Comment by Tyriar** [comment]
@bpasero it was definitely not there, only `compile` was in that dir. I can check the version tonight.

**Comment by bpasero** [comment]
Make sure its version is 0.11.0

**Comment by Tyriar** [comment]
It was 0.10.6, deleting `node_modules` and running `npm install` in `extensions/vscode-api-tests` fixed it. I'm not able to run `./scripts/npm.sh install --arch=x64` in the project root to completion.

Do you know what could have caused the module not installing correctly? Did it `vscode-api-tests` not bother to fetch the latest version because it wanted the `next` version or something?

**Comment by bpasero** [comment]
Looks like this works now. Somehow the `vscode: next` was resolving to `0.10.7` which does not include the post install script. I since released `vscode@0.11.0` and advanced the `next` tag and now cannot reproduce anymore.
